### PR TITLE
WIP: Unhealthy on storage blackout

### PR DESCRIPTION
**Pull request by iori-yja**
## Which problem is this PR solving?
The health status doesn't tell the healthiness of the server after initialization.
https://github.com/jaegertracing/jaeger/issues/1331#issuecomment-462670392

### Goal
- a part of the server can declare its _unhealthy_ state
- if one of essential parts declares unhealthy, then server is unhealthy
- configurable what is the essential part

### CODE REVIEW

**Comment by yurishkuro** [comment]
I am not sure this approach works. In particular, in Cassandra it only sets check-passed once the Session is created. But if Cassandra is unavailable, the factory will simply return an error and abort the application, so what does the health check actually provide?

I think the main change here is not so much connecting storage status to the health check, but actually _having storage status_. Right now the status has only one value: ready. If not ready - app dies.

**Comment by iori-yja** [comment]
What does "abort the application" mean? Actually, it don't die after the first connection succeeded. Particularly, if the application connected to a cassandra server _right after_ it has got available, the data gets never be available because connection pool is empty (or gone).

Possibly it can be fixed by adding another check routine in the factory, though...

**Comment by iori-yja** [comment]
> I am not sure this approach works. In particular, in Cassandra it only sets check-passed once the Session is created. But if Cassandra is unavailable, the factory will simply return an error and abort the application, so what does the health check actually provide?

I understand this point. I am working on propagating `StatusReporter` to reader/writer part of store (eg. spanstore), but this is a little bit messy.
Another approach is to make it just die after expiring `max-retry-attempts` which sounds more easy. But there must be pros/cons.

> If not ready - app dies.
I am going to add an option to specify which part is required to be "ready". If "Initialization" and "Storage" are required, storage must be ready and healthy for the application to be ready. If it is only "Initialization", the behavior is same as now.

**Comment by iori-yja** [comment]
Following DCO guide ruined the history....

**Comment by iori-yja** [comment]
Sorry, I said "never", but it turned out not true. Leaving long time (sorry, I've failed to measure) can solve the problem. So the problem only left here is the health status reports ok before the database gets ready.

File: plugin/storage/cassandra/factory.go [hunk 1]
```diff
@@ -71,10 +73,11 @@ func (f *Factory) InitFromViper(v *viper.Viper) {
...
 // Initialize implements storage.Factory
-func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger) error {
+func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger, reporter healthcheck.StatusReporter) error {
 	f.primaryMetricsFactory = metricsFactory.Namespace(metrics.NSOptions{Name: "cassandra", Tags: nil})
 	f.archiveMetricsFactory = metricsFactory.Namespace(metrics.NSOptions{Name: "cassandra-archive", Tags: nil})
 	f.logger = logger
+	f.StatusReporter = reporter
 
 	primarySession, err := f.primaryConfig.NewSession()
 	if err != nil {
```
Comment (yurishkuro) [inline plugin/storage/cassandra/factory.go]: if Cassandra is not ready, this will cause the application to exit

Comment (iori-yja) [reply plugin/storage/cassandra/factory.go]: Sorry, I took a while to move back on this.
 
I think my previous explanation was bad.
Actually, the symptom is triggered by cassandra restart. In that case, the session pool become empty after the initialization. In addition, gocql doesn't deal with the reconnection, which makes the jaeger services unavailable after that.

**Comment by jpkrohling** [comment]
@iori-yja are you still interested in this one?

**Pull request closed without merging**